### PR TITLE
ddl: add index acceleration introduce sysvars and config param

### PR DESCRIPTION
**Pull request by Benjamin2037**
### What problem does this PR solve?
Add sysvars and config parameter for add index acceleration.

Issue Number: ref #35983

Problem Summary:

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -741,6 +741,10 @@ const (
 	// TiDBMaxAutoAnalyzeTime is the max time that auto analyze can run. If auto analyze runs longer than the value, it
 	// will be killed. 0 indicates that there is no time limit.
 	TiDBMaxAutoAnalyzeTime = "tidb_max_auto_analyze_time"
+	// TiDBFastDDL indicates whether use lighting to help acceleate adding index stmt.
+	TiDBFastDDL = "tidb_fast_ddl"
```
Comment (tangenta) [inline sessionctx/variable/tidb_vars.go:745]: It is recommend to use `tidb_enable_fast_ddl` to make it consistent with other names.

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -741,6 +741,10 @@ const (
...
+	// TiDBDiskQuota used to set disk quota for lightning add index.
+	TiDBDiskQuota = "tidb_disk_quota"
```
Comment (tangenta) [inline sessionctx/variable/tidb_vars.go:747]: How about using `executor.GlobalDiskUsageTracker` to track the memory usage?

Comment (Benjamin2037) [reply sessionctx/variable/tidb_vars.go:747]: executor.GlobalDiskUsageTracker is for executor query check storage usage?
TiDBDiskQuota has two functions: 1) used to set a low bar to make sure there is at least disk space to enable lightning execution enviroment.  2) it also could let user change the lightning local disk quota to make add index lightning do partial ingest sst to TiKV. 
I think it has risk to mix merge TiDBDiskQuota into executor.GlobalDiskUsageTracker or could we delay this task into next stage if needed
Also I found the name is not very clear to show it meaning, then I will change it to TiDBDDLDiskQuota, is that ok?

File: sessionctx/variable/tidb_vars.go (hunk 1 above)
Comment (Benjamin2037) [reply sessionctx/variable/tidb_vars.go:745]: I will change it to tidb_enable_fast_ddl.

File: config/config.go [hunk 3]
```diff
@@ -85,6 +85,8 @@ const (
 	DefExpensiveQueryTimeThreshold = 60
 	// DefMemoryUsageAlarmRatio is the threshold triggering an alarm which the memory usage of tidb-server instance exceeds.
 	DefMemoryUsageAlarmRatio = 0.8
+	// DefFastReorgLocalPath is the default sort dir for add index lightning solution
```
Comment (tangenta) [inline config/config.go:88]:
```suggestion
	// DefFastReorgLocalPath is the default sort dir for add index with lightning solution.
```

File: config/config.go [hunk 4]
```diff
@@ -272,6 +274,8 @@ type Config struct {
 	CheckMb4ValueInUTF8        AtomicBool `toml:"check-mb4-value-in-utf8" json:"check-mb4-value-in-utf8"`
 	EnableCollectExecutionInfo bool       `toml:"enable-collect-execution-info" json:"enable-collect-execution-info"`
 	Plugin                     Plugin     `toml:"plugin" json:"plugin"`
+	// FastReorgLocalPath used to specific the lighting DDL local sort path.
```
Comment (tangenta) [inline config/config.go:277]:
```suggestion
	// FastReorgLocalPath is used to specify the lighting DDL local sort path.
```

File: sessionctx/variable/tidb_vars.go [hunk 5]
```diff
@@ -1044,6 +1051,10 @@ var (
 	PreparedPlanCacheMemoryGuardRatio = atomic.NewFloat64(DefTiDBPrepPlanCacheMemoryGuardRatio)
 	EnableConcurrentDDL               = atomic.NewBool(DefTiDBEnableConcurrentDDL)
 	EnableNoopVariables               = atomic.NewBool(DefTiDBEnableNoopVariables)
+	// EnableFastReorg indicates whether to use lightning to enhance DDL reorg performance.
+	EnableFastReorg = atomic.NewBool(false)
```
Comment (tangenta) [inline sessionctx/variable/tidb_vars.go:1055]:
```suggestion
	EnableFastReorg = atomic.NewBool(DefTiDBEnableFastReorg)
```

It is better to keep the style consistent with other variables.

File: sessionctx/variable/tidb_vars.go [hunk 6]
```diff
@@ -1044,6 +1051,10 @@ var (
...
+	// Temporary Variable for set dist quota for lightning add index, int type, GB as unit
```
Comment (tangenta) [inline sessionctx/variable/tidb_vars.go:1056]:
```suggestion
	// DDLDiskQuota is the temporary variable for set disk quota for lightning add index, int type, GB as unit.
```

File: sessionctx/variable/sysvar_test.go [hunk 7]
```diff
@@ -1069,3 +1069,79 @@ func TestDefaultMemoryDebugModeValue(t *testing.T) {
...
+	// Set to 1 Pb
+	err = mock.SetGlobalSysVar(TiDBDDLDiskQuota, strconv.FormatInt(pb, 10))
+	require.NoError(t, err)
+	val, err = mock.GetGlobalSysVar(TiDBDDLDiskQuota)
+	require.NoError(t, err)
+	require.Equal(t, strconv.FormatInt(pb, 10), val)
+
+	// MaxValue is 1 PB, set to 2 Pb is not allowed
+	err = mock.SetGlobalSysVar(TiDBDDLDiskQuota, strconv.FormatInt(2*pb, 10))
+	require.NoError(t, err)
```
Comment (tangenta) [inline sessionctx/variable/sysvar_test.go:1143]: It is really "not allowed"?

File: config/config.go (hunk 3 above)
Comment (Benjamin2037) [reply config/config.go]: Done

File: config/config.go (hunk 4 above)
Comment (Benjamin2037) [reply config/config.go]: Done

File: sessionctx/variable/tidb_vars.go (hunk 5 above)
Comment (Benjamin2037) [reply sessionctx/variable/tidb_vars.go]: Done

File: sessionctx/variable/tidb_vars.go (hunk 6 above)
Comment (Benjamin2037) [reply sessionctx/variable/tidb_vars.go]: Done

**Review by tangenta — APPROVED** [review]

File: config/config.go [hunk 8]
```diff
@@ -272,6 +274,8 @@ type Config struct {
 	CheckMb4ValueInUTF8        AtomicBool `toml:"check-mb4-value-in-utf8" json:"check-mb4-value-in-utf8"`
 	EnableCollectExecutionInfo bool       `toml:"enable-collect-execution-info" json:"enable-collect-execution-info"`
 	Plugin                     Plugin     `toml:"plugin" json:"plugin"`
+	// FastReorgLocalPath is used to specify the lighting DDL local sort path.
+	FastReorgLocalPath string `toml:"fast-reorg-local-path" json:"fast-reorg-local-path"`
```
Comment (xiongjiwei) [inline config/config.go:278]: should we add this into `config.example.toml`?

File: sessionctx/variable/sysvar.go [hunk 9]
```diff
@@ -1694,6 +1694,7 @@ var defaultSysVars = []*SysVar{
 			metrics.ToggleSimplifiedMode(TiDBOptOn(s))
 			return nil
 		}},
+
```
Comment (xiongjiwei) [inline sessionctx/variable/sysvar.go:1697]: remove

File: sessionctx/variable/sysvar.go [hunk 10]
```diff
@@ -1719,6 +1720,19 @@ var defaultSysVars = []*SysVar{
...
 			return nil
 		}},
+	{Scope: ScopeGlobal, Name: TiDBDDLEnableFastReorg, Value: BoolToOnOff(DefTiDBEnableFastReorg), Type: TypeBool, GetGlobal: func(sv *SessionVars) (string, error) {
+		return BoolToOnOff(EnableFastReorg.Load()), nil
+	}, SetGlobal: func(s *SessionVars, val string) error {
+		EnableFastReorg.Store(TiDBOptOn(val))
+		return nil
+	}},
+	// This system var is set disk quota for lightning sort dir, from 100 GB to 1PB.
+	{Scope: ScopeGlobal, Name: TiDBDDLDiskQuota, Value: strconv.Itoa(DefTiDBDDLDiskQuota), Type: TypeInt, MinValue: DefTiDBDDLDiskQuota, MaxValue: 1024 * 1024 * DefTiDBDDLDiskQuota / 100, GetGlobal: func(sv *SessionVars) (string, error) {
```
Comment (xiongjiwei) [inline sessionctx/variable/sysvar.go:1730]: it will be a big number for 100GB shows in bytes, can we use GB directly, like `set @@tidb_ddl_disk_quota=100`

Comment (tangenta) [reply sessionctx/variable/sysvar.go:1730]: The other options like `tidb_memory_quota_query`, `tidb_stats_cache_quota` and `tidb_memory_quota_apply_cache` are counted in bytes. Let's improve them systematically: https://github.com/pingcap/tidb/issues/17847.

**Review by xiongjiwei — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**